src/datastore/connection.ts

```diff
@@ -26,9 +26,11 @@ export async function sqlTransaction<T>(
   readOnly = true
 ): Promise<UnwrapPromiseArray<T>> {
   if ('savepoint' in sql) {
-    return sql.savepoint(callback);
+    // Only the `postgres.TransactionSql` type contains a `savepoint` method, which means we're
+    // already inside a SQL transaction.
+    return callback(sql) as UnwrapPromiseArray<T>;
   } else {
-    return sql.begin(readOnly ? 'READ ONLY' : '', callback);
+    return sql.begin(readOnly ? 'read only' : '', callback);
```

Comment:
IIUC, this helps prevent/fix situations like:
```ts
sqlTransaction(store.sql, scopedSqlA => {
  scopedSqlA.query(...);
  // some indirection and nested function calls later, an unintentional new sql transaction:
  sqlTransaction(store.sql, scopedSqlB => {
    scopedSqlB.query(...); // localSqlB will now be the same instance as scopedSqlA with this PR
  });
})
```
right?

If so, this seems like it would fix the nested transaction-in-transaction warnings 👍 


I'm curious if there's anything we could do to also prevent scenarios like:
```ts
sqlTransaction(store.sql, scopedSqlA => {
  scopedSqlA.query(...);
  // some indirection and nested function calls later, an unintentional out-of-transaction sql query:
  store.sql.query(...); // oops, using a different sql client instance from the pool inside of the transaction of another sql client instance.. deadlocks? socket data corruption? idk?
  });
})
```

Reply:
That's exactly right... I've been looking at options this morning for this and the only concrete idea I've had is to actually **get rid** of `this.sql` altogether.

It would take some refactoring, but by doing that we would open a `sql` object only when we first need to interface with `PgStore` and from there on all the functions inside in the call stack would only have access to their scoped `sql` object preventing these mis-uses.

What do you think?

Reply:
I think removing `this.sql` as an easy to use property makes sense. It's extremely easy to unintentionally reference it when you think you're using a transaction instance, or when you think you're not using one, or not knowing if you're already in one, etc..

What do you think? Something like requiring an explicit function with a callback in order to access to the `sql` client instance -- similar to the `sql.begin(...)` fn?

Reply:
Yeah exactly, that could work but we'd need to make sure it also works in a nested kind of way (so we don't run into the same problem i.e. a call to `getNewClient(sql ...)` inside one that already exists).

I'll try some options and let you know

Reply:
@zone117x I've pushed a sample of the code I'm currently working on to get your opinion.

In a nutshell, I've added a new `sqlSession` helper here: https://github.com/hirosystems/stacks-blockchain-api/pull/1376/commits/4ed98de8db081eaee672d14a653cd7e8a7ca96a6 that creates a new connection, runs a callback with that SQL object, and closes the connection. The helper also allows us to set a name per session which will make query errors much easier to debug, for example: https://github.com/hirosystems/stacks-blockchain-api/pull/1376/commits/4ed98de8db081eaee672d14a653cd7e8a7ca96a6. I've also removed `this.sql` from `PgStore`.

To absolutely prevent anyone from calling `sqlSession` in a nested fashion, we could add a custom eslint rule (maybe in another PR).

What do you think?

Reply:
How long do we expect the `sql` instance inside of the `sqlSession(...)` callback to last? It looks like it's intended usage to not last for very long (e.g. only for the duration of a given http query). If so, we'd be missing out on a lot of benefits provided by connection pooling, such as avoiding the renegotiating connections to pg constantly, and cached prepared statements (pretty sure this cache is associated with the connection lifetime).

Reply:
I'm thinking.. conceptually if this was another language and we used synchronous threads (where each session of sql queries would take place on the same thread), then it would be pretty simple to ensure a "thread local" value that could indicate if a sql transaction is open, and a helper function to open a sql transaction could check this thread local and throw.

I started looking around to see if there was anything like an "AsyncLocal" in js (it's conceptually similar but instead of variable persisted through a thread, it's a variable persisted through an asynchronous flow (a series of async/await calls). For example see C#'s [`AsyncLocal`](https://learn.microsoft.com/en-us/dotnet/api/system.threading.asynclocal-1?view=net-6.0).

Then I came across the nodejs API `async_hooks` which I haven't really looked at before. It appears to provide this "AsyncLocal" capability, and their docs even have example code referencing db usage with the API: https://nodejs.org/docs/latest-v16.x/api/async_context.html#usage-with-asyncawait

So it kinda looks like we could use (and potentially even without touching a ton of code) to enforce a given async flow never opens more than one sql transaction at once. What do you think?